runtime/storage/in-memory-storage.js

```diff
@@ -241,21 +270,34 @@ class InMemoryVariable extends InMemoryStorageProvider {
       let value = this._stored;
       let referredType = this.type.referenceReferredType;
       // TODO: string version of ReferredTyped as ID?
-      // TODO: remember to use/check the ID of the reference too.
-      let store = await this._storageEngine.connect(referredType.toString(), referredType, value.storageKey);
-      let result = await store.get();
+      if (this._backingStore == null)
+        this._backingStore = await this._storageEngine.share(referredType.toString(), referredType.collectionOf(), value.storageKey);
+      let result = await this._backingStore.get(value.id);
       return result;
     }
     return this._stored;
   }
 
   async set(value, originatorId=null, barrier=null) {
     assert(value !== undefined);
-    // If there's a barrier set, then the originating storage-proxy is expecting
-    // a result so we cannot suppress the event here.
-    if (JSON.stringify(this._stored) == JSON.stringify(value) && barrier == null)
-      return;
-    this._stored = value;
+    if (this.type.isReference) {
+      // If there's a barrier set, then the originating storage-proxy is expecting
+      // a result so we cannot suppress the event here.
+      if (this.__stored && this.__stored.id == value.id && barrier == null)
+        return;
+      
+      let referredType = this.type.referenceReferredType;
+      if (this._backingStore == null)
+        this._backingStore = await this._storageEngine.baseStorageFor(referredType);
```

Comment:
It seems like this could be de-duped to a common call (getBackingStore)... but that would have to be async, which would presumably force a yield point on every call, whereas repeating this code means it's only the first time. Is that better? Or is always yielding in the same pattern more desirable?

Reply:
Yields will be mainly observable as latency, I think, so it's probably better not to yield when not required.